Add metadata support to Snapcast media players

mill1000 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add metadata support to Snapcast media players.

Snapcast streams can contain metadata provided via a [stream plugin](https://github.com/badaix/snapcast/blob/develop/doc/json_rpc_api/stream_plugin.md) or natively.

Snapcast includes support for the following sources:
- Librespot via stream plugin
- Mopidy via stream plugin
- MPD via stream plugin
- Airplay natively

Additionally I have developed a stream plugin for UPnP renderers: https://github.com/mill1000/snapcast-upnp

Example Screenshots
UPnP Stream Plugin
![image](https://github.com/home-assistant/core/assets/6924622/bb276db4-d279-4e23-99ad-5087613552e6)

Airplay Native
![image](https://github.com/home-assistant/core/assets/6924622/49a6b188-a1d2-419b-8fc8-bb944b3ffa79)


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes # N/A
- This PR is related to issue: N/A
- Link to documentation pull request: N/A

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [N/A] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [N/A] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [N/A ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [N/A ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [N/A ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

mill1000 committed:
Refactor Snapcast client and group classes to use a common base class.

mill1000 committed:
Add metadata support to Snapcast media players

mill1000 committed:
Use saved server reference insead of pulling from global data

mill1000 committed:
Reorganize properties so they're closer to associated methods

mill1000 committed:
Fix typo in remove

mill1000 committed:
Revert "Use saved server reference insead of pulling from global data"

This reverts commit fca4360dc7d5b35330d2c9da2ebcc55d1150aab8.

mill1000 committed:
Fix constructor arguments in HomeAssistantSnapcast

luar123 commented:
That looks like a nice addition, thanks!
I will review and test it, but might take some time.

luar123 reviewed (commented):
I have added a few small comments, nothing critical. In general it would be much easier to review if you would split it in two PRs: one for the refactor and one for metadata support. Metadata support could be simplified by adding metadata properties to the python-snapcast client/group classes first but I think your solution is fine.

mill1000 commented:
Thanks for taking time to review. I'll check out the comments.

> I have added a few small comments, nothing critical. In general it would be much easier to review if you would split it in two PRs: one for the refactor and one for metadata support. Metadata support could be simplified by adding metadata properties to the python-snapcast client/group classes first but I think your solution is fine.

If you want me to split the PR in two that's no problem.

I am also working on another change to implement the grouping in line with this [architecture discussion](https://github.com/home-assistant/architecture/issues/364#issuecomment-625405413). It adds the [media player grouping services](https://developers.home-assistant.io/docs/core/entity/media-player#grouping-player-entities-together) and attributes. That PR would make the refactor here redundant as it eliminates the Group entities all together.

mill1000 commented on homeassistant/components/snapcast/media_player.py:
```diff
@@ -114,166 +119,233 @@ class SnapcastGroupDevice(MediaPlayerEntity):
+        self._server = server
         self._entry_id = entry_id
-        self._attr_unique_id = f"{GROUP_PREFIX}{uid_part}_{self._group.identifier}"
+        self._attr_media_content_type = MediaType.MUSIC
```
I understand where you're coming from as the stream could be a broadcast, notification, etc,

But
A. `MediaType.MUSIC` is required for `artist` and `album` metadata to be parsed.
B. Snapcast doesn't currently provide a means to specify this.

Of the available MediaType, `MUSIC` and `PODCAST` might be the only relevant types.

```python
class MediaType(StrEnum):
    """Media type for media player entities."""

    ALBUM = "album"
    APP = "app"
    APPS = "apps"
    ARTIST = "artist"
    CHANNEL = "channel"
    CHANNELS = "channels"
    COMPOSER = "composer"
    CONTRIBUTING_ARTIST = "contributing_artist"
    EPISODE = "episode"
    GAME = "game"
    GENRE = "genre"
    IMAGE = "image"
    MOVIE = "movie"
    MUSIC = "music"
    PLAYLIST = "playlist"
    PODCAST = "podcast"
    SEASON = "season"
    TRACK = "track"
    TVSHOW = "tvshow"
    URL = "url"
    VIDEO = "video"
```

We could attempt to deduce the type by checking the metadata. If an `album` or `artist` property is present set the type to `MUSIC` otherwise leave it empty.

luar123 replied:
Good point, didn't know that. For me both solutions are fine.
Maybe it is easier then to set it to Music as you suggest.

mill1000 replied:
Unless anyone else feels strongly I would prefer to leave this set to Music.

mill1000 committed:
Simplify media_artist and media_album_artist properties based on review

luar123 commented on homeassistant/components/snapcast/media_player.py:
```diff
@@ -114,166 +119,233 @@ class SnapcastGroupDevice(MediaPlayerEntity):
         self._entry_id = entry_id
-        self._attr_unique_id = f"{GROUP_PREFIX}{uid_part}_{self._group.identifier}"
+        self._attr_media_content_type = MediaType.MUSIC
+        self._attr_device_class = MediaPlayerDeviceClass.SPEAKER
```
I guess if there is no need to set it I would leave it unset. If there is a benefit of setting this to speaker we could also go that way, as it is the most common case.

mill1000 replied:
Understood. I'm in the middle of a move right now so I'm unable to determine if it's necessary or not.

luar123 commented:
> If you want me to split the PR in two that's no problem.

Honestly, I don't care, but core devs (need to approve and merge) sometimes prefer smaller PRs. You could try to ask on discord.

Offtopic:
> I am also working on another change to implement the grouping in line with this [architecture discussion](https://github.com/home-assistant/architecture/issues/364#issuecomment-625405413). It adds the [media player grouping services](https://developers.home-assistant.io/docs/core/entity/media-player#grouping-player-entities-together) and attributes. That PR would make the refactor here redundant as it eliminates the Group entities all together.

I would love to see that feature. Was just thinking of adding a option to configflow/option flow to still keep group entities for backwards compatibility. (I don't use them at all, but others might.)

luar123 commented on homeassistant/components/snapcast/media_player.py:
```diff
@@ -114,166 +119,227 @@ class SnapcastGroupDevice(MediaPlayerEntity):
 
-class SnapcastClientDevice(MediaPlayerEntity):
-    """Representation of a Snapcast client device."""
+        return None
```
return default

luar123 commented:
I tried it on my hass server. Works perfect after changing the return value of `_get_metadata()`.

Chreece commented:
That's a great feature to add. I just found this PR and wanted to test it by copying the snapcast folder from PR to my custom_components. 
I'm using mopidy as source for the snapcast, the metadata are viewable through webview (http:/<snapserver_ip>:1780) but still not in home assistant snapclients.  Am I missing something or mopidy isn't supported?

mill1000 commented:
> Is it also possible to add control (play-pause-next etc)?

 If the stream plugin supports control it could be possible, but I think I'd save it for another PR

luar123 commented:
> If the stream plugin supports control it could be possible, but I think I'd save it for another PR

I think first this functionality needs to be added to python-snapcast.

luar123 commented on homeassistant/components/snapcast/media_player.py:
```diff
@@ -114,166 +119,227 @@ class SnapcastGroupDevice(MediaPlayerEntity):
 
-class SnapcastClientDevice(MediaPlayerEntity):
-    """Representation of a Snapcast client device."""
+        return None
```
```suggestion
        return default
```

mill1000 replied:
Thanks for catching this

mill1000 committed:
Return default in get_metadata

mill1000 commented:
Fixed the `_get_metadata` return and rebased against the latest dev branch.

I think the only potentially unresolved comments are how to deal with the following.
```
  self._attr_media_content_type = MediaType.MUSIC
  self._attr_device_class = MediaPlayerDeviceClass.SPEAKER
```
And if there's reason not to hardcode them.

luar123 reviewed (approved):
Thanks, fine for me.
(CI crashed, you can either force push again or close/re-open the PR to trigger it again.)

emontnemery commented:
@mill1000 As previously proposed, would it be possible to split this PR up please?
- A PR which adds type annotations
- A PR which does the refactoring to a base class
- A PR which adds the metadata support

Such a set of smaller PR is *much* easier to review.

mill1000 commented:
Yeah no problem. I'll get to work on that today.

mill1000 commented:
I see that @frenck force-pushed and added the `smash` label a few months ago. 

Just want to confirm I'm not disrupting any plans by splitting this PR.

emontnemery commented:
> I see that @frenck force-pushed and added the `smash` label a few months ago.
> 
> Just want to confirm I'm not disrupting any plans by splitting this PR.

No, that won't disrupt any plans. Splitting the PR up is most welcome👍

mill1000 commented:
I'm a little uncertain of the value a PR just for type-annotations. Seems like all the changes made in a PR for that would be wiped out in the refactor PR.

luar123 commented:
Just do the refactor PR first then.

mill1000 commented:
Refactor PR here: https://github.com/home-assistant/core/pull/124499

Closing this original PR.
